sys/arvo.hoon

```diff
@@ -683,6 +689,83 @@
   ::
   [[[~ %vega hap] ((list ovum) -.out)] +.out]
 ::
+++  velo                                                ::  new full reboot
+  |=  $:  ::  now: current date
+          ::  ova: actions to process after reboot
+          ::  hun: hoon.hoon source
+          ::  arv: arvo.hoon source
+          ::
+          now=@da
+          ova=(list ovum)
+          hun=@t
+          van=@t
+      ==
+  ^-  (unit (pair (list ovum) *))
+  ::  virtualize; dump error if we fail
+  ::
+  =-  ?:(?=(%| -.res) ((slog p.res) ~) `p.res)
+  ^=  res  %-  mule  |.
+  ::  produce a new kernel and an effect list
+  ::
+  ^-  (pair (list ovum) *)
+  ::  compile the hoon.hoon source with the current compiler
+  ::
```

Comment:
Philosophically, should this process happen in the kernel?

If this was to happen in user space, we could do all sorts of validation before userspace handed arvo its new self.

Reply:
It's an interesting question -- to what degree is it appropriate for Arvo proper to defer to its modules? [neo arvo](https://github.com/cgyarvin/arvo/blob/remaint-merge/neo/arvo.hoon) (which will need a closer look and broader discussion at some point in the near future) goes in the opposite direction, giving Arvo a miniature filesystem and additional independence from the vanes. I'm personally torn, I like the idea of a more independent Arvo, but is it worth the maintenance and conceptual burden? Does Arvo ever need to reset itself without having vanes and a filesystem?

For now, I think it's clear that `%velo` is superior to `%vega` (of course, I implemented neither of these ...)

/cc @cgyarvin

Reply:
Apologies for the long-winded comment.

The idea of Arvo having a filesystem but not preloaded state in other vanes seems a bit hacky to me. A filesystem is just an initial Clay fixture. I feel like an initial Jael snapshot and an initial Ames map from ship to IP address should go in the same place.

If we go the route of Arvo handling its own updates, we might want to give it an argument of validation checks that it can perform -- things like unit tests. Unfortunately, making that fully general is quite difficult, since we might want it to be able to run arbitrary Ford builds and do other things that require asynchronicity or remote data, and that would be pretty ugly to build into Arvo at that level.,

For that reason, I think I'd be happier taking Arvo in a more minimal direction long-term, where maybe it just receives a vase of a new vane. That way we can separate the concern of constructing that vane from the concern of installing the vane, and we can build arbitrarily complex development logic for the developer workflow without corrupting the cleanliness of Arvo itself.

There are also potential performance implications based on where the construction of new vanes happens. If it happens in userspace, it can use Ford's caching to generate kernels faster, which could save a lot of time when doing kernel and vane development. Then again, long-term I suspect we'll want a persistent compiler cache in Vere, which should get a hit when rebuilding the same vane, no matter whether that happens through Ford or otherwise.

Reply:
Sorry, I think I was unclear. That new Arvo implementation doesn't have a %clay snapshot or something of that sort. It has its own "deep storage" -- what appears to be a small filesystem implementation to back kernel reboots and vane installations.

I agree that we will want the flexibility you've described. But I'm not sure that means Arvo needs to just accept already-built vanes as vases. That sounds kind of crippled to me. Definitely more to discuss here.